common/headers/headers.go

```diff
@@ -90,3 +93,38 @@ func (h GRPCHeaderGetter) Get(key string) string {
 	}
 	return ""
 }
+
+const (
+	// maxExperiments limits the number of experiments that can be specified in the header
+	// to prevent resource exhaustion from parsing arbitrarily large comma-separated lists.
+	maxExperiments = 10
+)
+
+// IsExperimentEnabled checks if a specific experiment is present in the temporal-experiment header.
+// Returns true if the experiment is explicitly listed or if "*" (wildcard) is present.
+// The header can contain a comma-separated list of up to maxExperiments experiment names.
+func IsExperimentEnabled(ctx context.Context, experiment string) bool {
+	experimentalValues := metadata.ValueFromIncomingContext(ctx, ExperimentHeaderName)
+
+	for _, headerValue := range experimentalValues {
+		if headerValue == "" {
+			continue
+		}
+		// Use SplitN to limit the number of parts we process, preventing
+		// resource exhaustion from maliciously crafted headers with excessive commas.
+		requestedExperiments := strings.SplitN(headerValue, ",", maxExperiments+1)
```

Comment:
nit: you could potentially save the split here and iterate on string indices but that's probably over optimization so you can ignore me :)